v3/mathjax3-ts/input/asciimath/FindAsciiMath.ts

```diff
@@ -0,0 +1,173 @@
+     */
+    protected start: RegExp;
+
+    /*
+     * The end-delimiter data keyed to the opening delimiter string
+     */
+    protected end: {[name: string]: ENDITEM};
+
+    /*
+     * False if the configuration has no delimiters (so search can be skipped), true otherwise
+     */
+    protected hasPatterns: boolean;
+
+    /*
+     * @override
+     */
+    constructor(options: OptionList) {
+        super(options);
+        this.getPatterns();
+    }
+
+    /*
+     * Create the patterns needed for searching the strings for AsciiMath
+     *   based on the configuration options
+     */
+    protected getPatterns() {
+        let options = this.options;
+        let starts: string[] = [];
+        this.end = {};
+        options['delimiters'].forEach((delims: DELIMS) => this.addPattern(starts, delims, false));
+        this.start = new RegExp(starts.join('|'), 'g');
+        this.hasPatterns = (starts.length > 0);
+    }
+
+    /*
+     * Add the needed patterns for a pair of delimiters
+     *
+     * @param{string[]} starts  Array of starting delimiter strings
+     * @param{DELIMS} delims    Array of delimiter strings, as [start, end]
+     * @param{boolean} display  True if the delimiters are for display mode
+     */
+    protected addPattern(starts: string[], delims: DELIMS, display: boolean) {
+        let [open, close] = delims;
+        starts.push(quotePattern(open));
+        this.end[open] = [close, display, new RegExp(quotePattern(close), 'g')];
+    }
+
+    /*
+     * Search for the end delimiter given the start delimiter.
+     *
+     * @param{string} text            The string being searched for the end delimiter
+     * @param{number} n               The index of the string being searched
+     * @param{RegExpExecArray} start  The result array from the start-delimiter search
+     * @param{ENDITEM} end            The end-delimiter data corresponding to the start delimiter
+     * @return{ProtoItem}             The proto math item for the math, if found
+     */
+    protected findEnd(text: string, n: number, start: RegExpExecArray, end: ENDITEM) {
+        let [close, display, pattern] = end;
+        let i = pattern.lastIndex = start.index + start[0].length;
+        let match = pattern.exec(text);
+        return (!match ? null : MATCH(start[0], text.substr(i, match.index - i), match[0],
+                                      n, start.index, match.index + match[0].length, display));
+    }
+
+    /*
+     * Search a string for math delimited by one of the delimiter pairs.
+     *
+     * @param{ProtoItem[]} math  The array of proto math items located so far
+     * @param{number} n          The index of the string being searched
+     * @param{string} text       The string being searched
+     */
+    protected findMathInString(math: ProtoItem[], n: number, text: string) {
+        let start, match;
+        this.start.lastIndex = 0;
+        while ((start = this.start.exec(text))) {
```

Comment:
Without these, some browsers (Firefox or Chrome, I forget which), give a warning about this, since it might look like a typo for `while (start == this.start.exec(text)) {` with `==` instead of `=`.  The extra parenthesis are what it says to use to prevent the warning.